tests/api/SubmissionsTest.php

```diff
@@ -0,0 +1,182 @@
+ *
+ * @class SubmissionsTest
+ * @ingroup tests_api
+ * @see SubmissionHandler
+ *
+ * @brief Test class for the /submissions endpoint
+ */
+
+import('lib.pkp.tests.PKPApiTestCase');
+
+class SubmissionsTest extends PKPApiTestCase {
+	/**
+	 * @covers /submissions
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionsWithoutToken() {
+		$response = $this->_sendRequest('GET', '/submissions', array(), false);
+	}
+
+	/**
+	 * @covers /submissions
+	 */
+	public function testGetSubmissions() {
+		$response = $this->_sendRequest('GET', '/submissions');
+		$this->assertEquals(200, $response->getStatusCode());
+		
+		$data = $this->_getResponseData($response);
+		$this->assertArrayHasKey('itemsMax', $data);
+		$this->assertArrayHasKey('items', $data);
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}/galleys
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionGalleysWithoutToken() {
+		$response = $this->_sendRequest('GET', '/submissions/9/galleys', array(), false);
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}/galleys
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionGalleysWithInvalidId() {
+		$response = $this->_sendRequest('GET', '/submissions/999/galleys');
+		$this->assertEquals(404, $response->getStatusCode());
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}/galleys
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionGalleysForUnpublishedSubmissions() {
+		$response = $this->_sendRequest('GET', '/submissions/23/galleys');
+		$this->assertEquals(404, $response->getStatusCode());
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}/galleys
+	 */
+	public function testGetSubmissionGalleys() {
+		$response = $this->_sendRequest('GET', '/submissions/9/galleys');
+		$this->assertEquals(200, $response->getStatusCode());
+		$data = $this->_getResponseData($response);
+		$this->assertEquals(1, count($data));
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionByIdWithValidIdWithoutToken() {
+		$response = $this->_sendRequest('GET', '/submissions/25', array(), false);
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}
+	 * @expectedException GuzzleHttp\Exception\ClientException
+	 */
+	public function testGetSubmissionByIdWithInvalidId() {
+		$response = $this->_sendRequest('GET', '/submissions/9999');
+		$this->assertSame(404, $response->getStatusCode());
+	}
+
+	/**
+	 * @covers /submissions/{submissionId}
+	 */
+	public function testGetSubmissionByIdWithValidId() {
+		$response = $this->_sendRequest('GET', '/submissions/25');
+		$this->assertSame(200, $response->getStatusCode());
+
+		$data = $this->_getResponseData($response);
+		$this->assertArrayHasKey('id', $data);
```

Comment:
These kinds of assertions start to duplicate the Swagger document. I wonder whether something like https://github.com/nabbar/SwaggerValidator-PHP wouldn't allow us to write tests here but still delegate the validation (e.g. required fields) to the Swagger schema.

Reply:
I'm afraid not -- I haven't worked with any of these tools. But something that validates a response against a Swagger description should be possible -- that link above is just a suggestion.

Reply:
I believe Swagger switched from JSON to YAML a while back and it looks like some of the tooling hasn't caught up. A couple of options come to mind:

1. Parse YAML into PHP and convert to JSON. http://php.net/manual/en/function.yaml-parse.php Will that work?

2. Parse YAML into PHP and roll our own tools to check API responses against valid properties and variable types.

3. Check out one of the other libraries: https://github.com/Maks3w/SwaggerAssertions or https://github.com/byjg/php-swagger-test

It looks like Postman has a nice toolset for creating, testing and documenting APIs. But it's all hosted in "the cloud", and I think we're on the right track with keeping the documentation under version control in the source. Apparently we can export from Postman to a JSON file. If this happened to match the Swagger style, that could work well.

I haven't found a nice, well-maintained technical-spec-to-user-documentation tool like Swagger's. It seems like we are going to have to close a gap in tooling somewhere. It's probably easiest to do that on the testing side, reading from the YAML (or JSON) and automating the checks.